src/prefect/workers/base.py

```diff
@@ -1562,6 +1591,267 @@ async def _mark_flow_run_as_cancelled(
+        try:
+            await self._poll_for_cancelling_flow_runs()
+        except Exception as exc:
+            self._logger.warning(
+                "Failed to poll for cancelling flow runs on startup: %s", exc
+            )
+
+        # Start the WebSocket consumer task
+        self._cancellation_consumer_task = asyncio.create_task(
+            self._consume_cancellation_events(),
+            name="cancellation-event-consumer",
+        )
+        self._cancellation_consumer_task.add_done_callback(
+            self._handle_cancellation_consumer_done
+        )
+
+    def _handle_cancellation_consumer_done(self, task: asyncio.Task[None]) -> None:
+        """Handle when the cancellation consumer task completes or fails.
+
+        If the WebSocket consumer fails, fall back to polling for cancelling flow runs.
+        """
+        if task.cancelled():
+            return
+        if exc := task.exception():
+            poll_seconds = get_current_settings().worker.cancellation_poll_seconds
+            self._logger.warning(
+                "Cancellation event consumer failed. Falling back to polling every %s seconds.",
+                poll_seconds,
+                exc_info=exc,
+            )
+            self._cancellation_polling_task = asyncio.create_task(
+                critical_service_loop(
+                    workload=self._poll_for_cancelling_flow_runs,
+                    interval=poll_seconds,
+                    jitter_range=0.3,
+                ),
+                name="cancellation-polling",
+            )
+            self._cancellation_polling_task.add_done_callback(
+                self._handle_cancellation_polling_done
+            )
+
+    def _handle_cancellation_polling_done(self, task: asyncio.Task[None]) -> None:
+        """Handle when the cancellation polling task completes or fails."""
+        if task.cancelled():
+            return
+        if exc := task.exception():
+            self._logger.error(
+                "Cancellation polling task failed. Pending flow runs will not be "
+                "cancelled automatically until the worker is restarted.",
+                exc_info=exc,
+            )
+
+    async def _poll_for_cancelling_flow_runs(self) -> None:
+        """Poll for cancelling flow runs in this work pool."""
+        cancelling_flow_runs = await self.client.read_flow_runs(
+            flow_run_filter=FlowRunFilter(
+                state=FlowRunFilterState(
+                    type=FlowRunFilterStateType(any_=[StateType.CANCELLING]),
+                ),
+            ),
+            work_pool_filter=WorkPoolFilter(
+                name=WorkPoolFilterName(any_=[self._work_pool_name])
+            ),
+        )
+
+        for flow_run in cancelling_flow_runs:
+            # Skip if we've already processed this flow run
+            if flow_run.id in self._seen_cancelling_flow_run_ids:
+                continue
+            self._seen_cancelling_flow_run_ids.add(flow_run.id)
```

Comment:
It looks like the worker only adds IDs to `_seen_cancelling_flow_run_ids`, but never removes them. Should there be something to trigger a removal or a TTL of some sort to avoid unbounded growth and/or allow for another cancellation attempt if a previous attempt fails?